File: marstv-backend/tests/Domain/DeviceChallengeTest.php
<?php
declare(strict_types=1);

namespace MarsTv\Tests\Domain;

use MarsTv\Domain\DeviceChallenge;
use PHPUnit\Framework\TestCase;

final class DeviceChallengeTest extends TestCase
{
    public function testExactSignedBytes(): void
    {
        $challenge = new DeviceChallenge('id', 'nonce', 'device', 'POST', '/api/v1/activation-sessions', 'hash', 123);
        self::assertSame("MARSTV_DEVICE_AUTH_V1\nid\nnonce\ndevice\nPOST\n/api/v1/activation-sessions\nhash\n123", $challenge->signedBytes());
    }
}

File: app/src/test/java/tv/mars/app/entitlement/DeviceChallengeTest.kt
package tv.mars.app.entitlement

import org.junit.Assert.assertArrayEquals
import org.junit.Assert.assertFalse
import org.junit.Test
import java.nio.charset.StandardCharsets

class DeviceChallengeTest {
    @Test
    fun `challenge bytes use exact line order without trailing newline`() {
        val challenge = DeviceChallenge("challenge", "nonce", "device", "POST", "/api/v1/status", "body", 123)
        val expected = "MARSTV_DEVICE_AUTH_V1\nchallenge\nnonce\ndevice\nPOST\n/api/v1/status\nbody\n123"

        assertArrayEquals(expected.toByteArray(StandardCharsets.UTF_8), challenge.signingBytes())
        assertFalse(challenge.signingBytes().last() == '\n'.code.toByte())
    }

    @Test(expected = IllegalArgumentException::class)
    fun `challenge rejects non-canonical path`() {
        DeviceChallenge("challenge", "nonce", "device", "POST", "https://evil.test/status", "body", 123)
    }
}
